src/Codec/architecture.md

```diff
@@ -0,0 +1,71 @@
+# Converting BSON data through codecs
+
+Codecs provide a more flexible way to convert BSON data to native types and back and address most of the shortcomings of
+the previous type map system. The codec system is designed to be used by libraries that need to convert BSON data into
+native types, for example object mappers. Unlike the type map system, codecs allow converting any BSON type to a native
+type directly when reading data from the database. Together with lazy decoding of BSON structures, this allows
+for a more flexible and efficient way to handle BSON data.
+
+## Encoders and Decoders
+
+The `Codec` interface is comprised of two smaller interfaces: `Decoder` and `Encoder`. The interfaces are typed as
+generics through `@template` annotations, allowing for better type checking when they are used.
+
+Each decoder and encoder may support a limited set of types or even values. Before calling `decode` or `encode`, the
+`canDecode` and `canEncode` methods can be used to determine whether a value is supported. If a value is not supported,
+the `decode` and `encode` methods will throw a `UnsupportedValueException` exception. The `decodeIfSupported` and
+`encodeIfSupported` methods are useful to encode or decode a value only if it is supported. For unsupported values, the
+original value is returned.
+
+## Codec Interfaces
+
+The `Codec` interface combines decoding and encoding into a single interface. This will be used for most values except
+for documents where a more specific `DocumentCodec` is provided. The `DocumentCodec` interface overrides the `decode`
+and `encode` methods from the base `Codec` interface to narrow the return types. Document codecs guarantee to always
+encode to a BSON document instance and decode to a PHP object.
+
+## Built-in Codecs
+
+By default, two codecs are provided: an `ArrayCodec` and an `ObjectCodec`. These two codecs are used to recursively
```

Comment:
@alcaeus this classes doesn't exist.

Reply:
Thanks for pointing this out. Removed in https://github.com/mongodb/mongo-php-library/pull/1181.